src/test/scala/com/mozilla/spark/sql/hyperloglog/test/HyperLogLog.scala

```diff
@@ -15,31 +15,83 @@ package com.mozilla.spark.sql.hyperloglog.test
 
 import com.mozilla.spark.sql.hyperloglog.aggregates._
 import com.mozilla.spark.sql.hyperloglog.functions._
-import org.apache.spark.sql.SQLContext
-import org.apache.spark.sql.functions._
-import org.apache.spark.{SparkConf, SparkContext}
-import org.scalatest.{FlatSpec, Matchers}
-
-class HyperLogLogTest extends FlatSpec with Matchers{
- "Algebird's HyperLogLog" can "be used from Spark" in {
-  val sparkConf = new SparkConf().setAppName("HyperLogLog")
-  sparkConf.setMaster(sparkConf.get("spark.master", "local[1]"))
-
-  val sc = new SparkContext(sparkConf)
-  val sqlContext = new SQLContext(sc)
-  import sqlContext.implicits._
-
-  val hllMerge = new HyperLogLogMerge
-  sqlContext.udf.register("hll_merge", hllMerge)
-  sqlContext.udf.register("hll_create", hllCreate _)
-  sqlContext.udf.register("hll_cardinality", hllCardinality _)
-
-  val frame = sc.parallelize(List("a", "b", "c", "c"), 4).toDF("id")
-  val count = frame
-    .select(expr("hll_create(id, 12) as hll"))
-    .groupBy()
-    .agg(expr("hll_cardinality(hll_merge(hll)) as count"))
-    .collect()
-  count(0)(0) should be (3)
- }
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.functions.expr
+import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
+
+case class ClientRow(id: String, os: String)
+
+class HyperLogLogTest extends FlatSpec with Matchers with BeforeAndAfterAll {
+
+  val spark = SparkSession.builder()
+      .master("local[*]")
+      .appName("Spark HyperLogLog Test")
+      .getOrCreate()
+
+  val osList = "windows" :: "macos" :: "linux" :: Nil
+  val predata = List(
+    ClientRow("a", "windows"),
+    ClientRow("b", "windows"),
+    ClientRow("c", "macos"),
+    ClientRow("c", "macos"),
+    ClientRow(null, "linux")
+  )
+
+  "Algebird's HyperLogLog" can "be used from Spark" in {
+    import spark.implicits._
+
+    val hllMerge = new HyperLogLogMerge
+    spark.udf.register("hll_merge", hllMerge)
+    spark.udf.register("hll_create", hllCreate _)
+    spark.udf.register("hll_cardinality", hllCardinality _)
```

Comment:
Can you register these UDFs once in a setup function rather than within each test?